### ISSUE TITLE
Binding for co-log

### ISSUE THREAD

**Issue by mmhat** [Kleidukos/effectful-contrib#10]
[co-log](https://hackage.haskell.org/package/co-log)

**Comment by ambroslins** [comment]
Hi, I would like to work on this issue, but I am stuck on choosing the right Monad `m` for the `LogAction m msg`.

After some trying around I came up with this: 
```haskell
data Log :: (Type -> Type) -> Type -> Effect

type instance DispatchOf (Log m msg) = 'Static

data instance StaticRep (Log m msg) = Log (LogAction m msg)

runLog :: LogAction (Eff es) msg -> Eff (Log (Eff es) msg ': es) a -> Eff es a
runLog = evalStaticRep . Log

logMsg :: Log (Eff es) msg :> es => msg -> Eff es ()
logMsg msg = do
  Log (LogAction action) <- getStaticRep
  action msg
```

However this does not work because the `Log` effect will always contain itself.
I think for this to work a subset like typeclass is required, like this:
```haskell
logMsg :: (Log (Eff logEffects) msg :> allEffects, Subset logEffects allEffects) => msg -> Eff allEffects ()
```

Another way i can think of is just fixing the Monad to `IO` (or `Eff [IOE]`),  like:
```haskell
data Log :: Type -> Effect
data instance StaticRep (Log  msg) = Log (LogAction IO msg)
```

I think, if the `Reader` effect would provide a `MonadReader` instance, most functions from co-log would just work with `WithLog`.

Basically my question is: what is the best way to move forward?

**Comment by ambroslins** [comment]
Looks like this will be solved soon: https://github.com/arybczak/effectful/issues/52

I will use this new class as the `Subset` class I proposed.

**Comment by arybczak** [comment]
I don't think it's going to work as you'd expect it to. 

I'm not familiar with `co-log`, but from what I see from briefly looking at hackage docs you should carry `LogAction IO msg` and execute it in `logMsg` using `unsafeEff`.

Also the StaticRep instance should be a `newtype`, not a `data`.

**Comment by ambroslins** [comment]
Thanks for the feedback.
You are right, this does not work as I expected, the `logEffects` would be ambiguous.
